dbms/src/Flash/Disaggregated/PageTunnel.cpp

```diff
@@ -0,0 +1,141 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/Exception.h>
+#include <Flash/Coprocessor/CHBlockChunkCodec.h>
+#include <Flash/Disaggregated/PageTunnel.h>
+#include <Interpreters/Context.h>
+#include <Storages/DeltaMerge/Delta/DeltaValueSpace.h>
+#include <Storages/DeltaMerge/DeltaMergeDefines.h>
+#include <Storages/DeltaMerge/Remote/DisaggSnapshot.h>
+#include <Storages/DeltaMerge/Remote/DisaggSnapshotManager.h>
+#include <Storages/DeltaMerge/Remote/Proto/remote.pb.h>
+#include <Storages/DeltaMerge/Segment.h>
+#include <Storages/Page/PageUtil.h>
+#include <Storages/Transaction/TMTContext.h>
+#include <common/logger_useful.h>
+#include <kvproto/mpp.pb.h>
+#include <tipb/expression.pb.h>
+
+#include <memory>
+
+namespace DB
+{
+PageTunnelPtr PageTunnel::build(
+    const Context & context,
+    const DM::DisaggTaskId & task_id,
+    TableID table_id,
+    UInt64 segment_id,
+    const PageIdU64s & read_page_ids)
+{
+    auto & tmt = context.getTMTContext();
+    auto * snap_manager = tmt.getDisaggSnapshotManager();
+    auto snap = snap_manager->getSnapshot(task_id);
+    RUNTIME_CHECK_MSG(snap != nullptr, "Can not find disaggregated task, task_id={}", task_id);
+    auto task = snap->popSegTask(table_id, segment_id);
+    RUNTIME_CHECK(task.isValid(), task.err_msg);
+
+    auto tunnel = std::make_unique<PageTunnel>(
+        task_id,
+        snap_manager,
+        task.seg_task,
+        task.column_defines,
+        task.output_field_types,
+        read_page_ids);
+    return tunnel;
+}
+
+disaggregated::PagesPacket PageTunnel::readPacket()
+{
+    // TODO: the returned rows should respect max_rows_per_chunk
+
+    disaggregated::PagesPacket packet;
+
+    // read page data by page_ids
+    size_t total_pages_data_size = 0;
+    auto persisted_cf = seg_task->read_snapshot->delta->getPersistedFileSetSnapshot();
+    for (const auto page_id : read_page_ids)
+    {
+        // FIXME: Depends on abstraction of reading page data from PageStorage directly
+        UNUSED(page_id);
+#if 0
+        auto page = persisted_cf->getStorage()->readForColumnFileTiny(page_id);
+        dtpb::RemotePage remote_page;
+        remote_page.set_page_id(page_id);
+        remote_page.mutable_data()->assign(page.data.begin(), page.data.end());
+        const auto field_sizes = PageUtil::getFieldSizes(page.field_offsets, page.data.size());
+        for (const auto field_sz : field_sizes)
+        {
+            remote_page.add_field_sizes(field_sz);
+        }
+        total_pages_data_size += page.data.size();
+        packet.mutable_pages()->Add(remote_page.SerializeAsString());
+#endif
```

Comment:
This piece of code depends on reading page data from SegmentSnapshot's PageStorage directly @breezewish

Reply:
resolved